### ISSUE TITLE
Web feature: key generation and decryption

### ISSUE THREAD

**Issue by balohmatevz** [kaepora/miniLock#185]
Greetings! 

I am a happy user of MiniLock, however there is one massive problem... not everyone uses it. While it is extremely good for communicating between my co-workers, it isn't quite as good when trying to securely send files to less-tech-savvy people. Think of situations where you have to email personal information to hotel receptionists, accountants, secretaries, HR workers and other potentially non-tech-savvy people, who do not offer you a secure way of doing this. 

What I would like to see on MiniLock's website is a feature I could easily link to in an email to the hypothetical hotel receptionist. When they click the link, it presents them with two fields fields:
- Recipient email address
- Password

They enter my email address and a password of their choosing. The password generation process can work exactly the same way as in the MiniLock Chrome app.

After they do this and hit 'send', a public key is created from the password they entered, which is then sent to my email address. I use it with MiniLock to encrypt the files and send them over email - the way they are used to receiving files. Once they receive these files, they go back to the MiniLock site, select the files I sent, enter the same password from earlier and get the files decrypted.

I usually find that non-tech-savvy people who work with personal information recognize that it is their job to keep it secure and tend to ensure privacy is well catered for when dealing with files on their computer or printed files. They however don't understand that email is insecure and that the 's' in 'https' actually matters. A service like this would increase the usefulness of MiniLock a lot, as it would allow us - as senders - to keep our data protected in transit even when sending to people who don't normally offer secure methods for data delivery. 

It is paramount to ensure that someone who doesn't know anything about cryptography can use the feature with as little mental strain as possible and as little extra time required on their part as possible. I also don't think it's extremely important that they know why they are doing this past "because a customer wanted it". So minimizing the amount of text might also be a good idea. 

This is follow up to question 4 on Security Now podcast #513

**Comment by SparkDustJoe** [comment]
Have you looked at Peerio.com? [Reply from my cell phone]

**Comment by balohmatevz** [comment]
I took a look at peerio.com and couldn't find anything I could easily link to someone to have them create a public key or decrypt files on the site. Imagine you are the person trying to get me to generate a public key for you and I have no clue what I'm doing, which link would you send to get me to generate a public key which you could use to encrypt the file you want to send me?

**Comment by SparkDustJoe** [comment]
You might need something like Zixmail.

**Comment by balohmatevz** [comment]
Actually... no... What I need is MiniLock, except available on the web - not as a Chrome extension - and with wording changed so it's geared towards people who don't speak geek... I took a look at how minilock's front end is implemented and realized it's trivial to deploy it as a website. After this I started playing around with the user interface and created a prototype of a user interface, suitable for recipients who don't know how to do encryption. It isn't perfect - I only spent a few hours on it - but it illustrates my idea. 

Unfortunately the current implementation doesn't work in Explorer and Safari. I didn't yet look whether the reason is UI related or crypto library related... 

You can try it here: http://errorage.com/ml/

**Comment by SparkDustJoe** [comment]
I'm beginning to understand more of what you're looking for based on the interface of that site you shared and re-reading your initial message.   I have a few points to make (NOTE: PERSONAL OPINIONS AHEAD):

1.  While I like the idea of having this be available to more people by having it publicly shared, there is an inherent danger in trusting code that runs from a browser versus the sandbox of Chrome's App model (which is precisely why Nadim chose that model versus hosting it on a web page).  You are trusting that the hosting site is using the correct source codes AND that no one is doing DNS spoofing or injection along the way to supplant any of the source files to make miniLock work in transit (which is an advanced and somewhat paranoid threat model, but one which is valid in modern times as demonstrated by NSA and GCHQ).   You are also trusting that other pages that you have visited have not installed some sort of toolbar or other malware on your machine that would override or supplant the JavaScript files running on the sender's machine/browser, since JavaScript is a very fluid and open-ended language.  The Chrome App model specifically protects the sources of the downloaded apps from tampering using a package/signature mechanism AND an isolated memory space that means other JavaScript or code cannot interfere with the running app, and that any memory used is isolated to the app (with the exception of some larger malware running on the sender's machine, like a key logger, which would defeat ANY secure system anyways and would be outside the scope of protection we're trying to achieve).  By downloading the official app from Google, and using it in that app model, it is walled off to JUST the sender's machine with a known set of source files.  A public website offers none of those advantages.

2.   By simplifying the key-pair generation mechanism, it does make the interface and overall use much more user friendly, BUT at the cost of security, as the restrictions on weak or common passwords is are removed.  This opens up the scheme to dictionary attacks as the space of possible source material for the system is drastically reduced.  Nadim specifically wanted to make attacker's lives difficult by using source material that had depth (more than one piece of material) AND complexity (>100 bit entropy input), where as your scheme does not. 

3.  As a continuation to the discussion above (in point 1), there is no guarantee that the website isn't secretly skimming the user's pass phrase or a copy of the sender's file, or injecting a 3rd party recipient into the header, if it requires a net connection to get to the service, whereas the Chrome App will function without net connectivity (and does not require it other than to download it in the first place, and subsequently to send/receive the files that have already been processed by the app offline).

So while I applaud the effort and I fully understand the user experience you have attempted to achieve, I have serious concerns (as the Devil's Advocate) for the security of the model.  There are trade-offs that you are making which tip the security scale a bit too far (in my opinion).

**Comment by SparkDustJoe** [comment]
Having said that, don't let my pessimism stop you, I think something like this has been long lacking in the world, and I've had to send forms with sensitive data in them to all kinds of people in far off locations throughout my professional career, and for me the solution ended up to being a paid customer for DropBox so that I could password protect special links to share files (that expire after 7 days).  Not ideal for everyone, especially for one-off forms.

**Comment by balohmatevz** [comment]
Yeah, I fully recognize this is an inferior product to the MiniLock app in terms of absolute security. However, you have to look at what the alternative is in practice: completely unencrypted email. The simplest of encryptions is better than plain text. Sure, it won't stop a targeted attack and won't stop state scale attacks, but it will stop just idle snooping. Of course it can make the sender complacent too by making them think they have encryption and can therefore do whatever, without understanding the degree to which encryption can protect them in practice...

In regard to common passwords and entropy, I have not looked at the source code for that, however the only difference in the source code related to cryptography is that the email address is not appended to the secret key and not included in entropy calculation. So password generation and entropy calculation should work the same as in the app. Again, I did not look at the crypto source code while doing this UI change, so I don't know whether this part does not work in a browser by default or if I broke something.

Everything this page does while working normally: creating the public key, encryption and decryption is done client-side. The file (encrypted or unencrypted) never leaves the browser, nor does the private key. The private key is stored in a javascript variable tho, so yes, malicious extensions with page content access permissions could get it or the files. At the moment, this only encrypts in a way that ensures only the holder of the private key can read it, there is no assurance of who the sender is, so a targeted attack could also replace the encrypted package while it's being emailed to the recipient, since all you need to encrypt it is the public key. 

So yeah... I recognize there are imperfections and targeted attacks will likely be able to work. It is however still far, far better than plain text. Think of this as a first iteration, a proof of concept, a demonstration of the idea - encryption that is "good enough for most uses" :D There is one more important aspect to this: education. Most people who work with private data (like doctors) fully understand it is private and that disclosing it can get them in serious trouble. However they don't even consider whether data in transit or at rest is secure - they simply assume it is. If customers begin to ask them to use a cryptography tool which they can intuitively do and understand, they will begin to see that the safety of data in transit is important too. Over time (as in over the course of decades), the understanding that the protection of data in transit is important will become something obvious to them, similar to how the immorality of disclosing data is obvious to them today.

**Comment by fruiz500** [comment]
I agree 100% with balohmatevz. If the app does not need to communicate with a server after being downloaded, then it can be made almost as secure as a Chrome app (which forces you to trust Google, anyway). The real problem to deal with is malware on your browser.

Concerning key length, an alternative would be to use variable key stretching, more stretching for lower entropy. This way low-entropy keys would remain available choices though users would be discouraged from using them, and hackers would still have to check them, at a considerable expense.
